File: medium/1081-smallest-subsequence-of-distinct-characters/solution.py
class Solution:
    def smallestSubsequence(self, s: str) -> str:
        """
        1081. Smallest Subsequence of Distinct Characters
        Time: O(n)
        Space: O(26)
        """
        last = {c: i for i, c in enumerate(s)}  # last occurrence of each char
        stack = []
        in_stack = set()

        for i, c in enumerate(s):
            if c in in_stack:
                continue
            # drop bigger characters that we can still re-add later
            while stack and stack[-1] > c and last[stack[-1]] > i:
                in_stack.remove(stack.pop())
            stack.append(c)
            in_stack.add(c)

        return "".join(stack)

File: medium/1081-smallest-subsequence-of-distinct-characters/solution.java
/**
 * 1081. Smallest Subsequence of Distinct Characters
 * Time: O(n)
 * Space: O(26)
 */
class Solution {
    public String smallestSubsequence(String s) {
        int[] last = new int[26]; // last occurrence of each char
        for (int i = 0; i < s.length(); i++) {
            last[s.charAt(i) - 'a'] = i;
        }

        StringBuilder stack = new StringBuilder();
        boolean[] inStack = new boolean[26];

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (inStack[c - 'a']) {
                continue;
            }
            // drop bigger characters that we can still re-add later
            while (stack.length() > 0) {
                char top = stack.charAt(stack.length() - 1);
                if (top > c && last[top - 'a'] > i) {
                    stack.deleteCharAt(stack.length() - 1);
                    inStack[top - 'a'] = false;
                } else {
                    break;
                }
            }
            stack.append(c);
            inStack[c - 'a'] = true;
        }

        return stack.toString();
    }
}
